src/mesh_tools.rs

```diff
@@ -5,8 +5,158 @@ use nalgebra::geometry::Point3;
+        .faces()
+        .iter()
+        .map(|face| match face {
+            Face::Triangle(f) => *f,
+        })
+        .collect();
+
+    // item index = face index; TRUE = the face was already checked, FALSE = the
+    // face hasn't yet been checked
+    let mut face_treatment_pattern = vec![false; original_triangle_faces.len()];
+
+    // Faces to be checked for winding, determined by the orientation of the
+    // OrientedEdge of the neighbor, which triggered the check. Current face has
+    // to contain a reverted edge to have the proper winding, otherwise it has
+    // to be reverted.
+    let mut edge_face_stack: Vec<(OrientedEdge, u32)> =
+        Vec::with_capacity(original_triangle_faces.len() / 2);
+
+    // Faces already checked and reverted if needed => all the faces in this
+    // list have the same vertex winding.
+    let mut synchronized_triangle_faces: Vec<TriangleFace> =
+        Vec::with_capacity(original_triangle_faces.len());
+
+    // Synchronize faces in all mesh geometry triangles
+    while synchronized_triangle_faces.len() < original_triangle_faces.len() {
+        // Start with the first untreated face in the list of the original mesh
+        // geometry faces. The winding of this face also determines the winding
+        // of the rest of the mesh geometry.
+        let mut current_face_index: u32 = 0;
+        while face_treatment_pattern[cast_usize(current_face_index)] {
+            current_face_index += 1;
+        }
+
+        face_treatment_pattern[cast_usize(current_face_index)] = true;
+        let mut current_test_edge: OrientedEdge =
+            original_triangle_faces[cast_usize(current_face_index)].to_oriented_edges()[0];
+
+        // Put the first edge-face couple to the stack
+        edge_face_stack.push((current_test_edge, current_face_index));
+
+        // Check and revert (if needed) faces in the stack as long as there is any
+        while !edge_face_stack.is_empty() {
+            // Get ready for the next iteration.
+            let next_edge_face = edge_face_stack.pop().expect("Popping from an empty vector");
+            current_test_edge = next_edge_face.0;
+            current_face_index = next_edge_face.1;
+
+            let current_face = original_triangle_faces[cast_usize(current_face_index)];
+            // Use if contains reverted edge, if not, revert and use. It should be
+            // safe to presume the face contains the edge one way or another because
+            // it's given by the topology generator.
+            let proper_current_face = if current_face.contains_oriented_edge(current_test_edge) {
+                current_face
+            } else {
+                current_face.to_reverted()
+            };
+            synchronized_triangle_faces.push(proper_current_face);
+
+            // Edge-to-index map for faster lookup
+            let mut unoriented_edge_index_map: HashMap<UnorientedEdge, u32> = HashMap::new();
```

Comment:
You don't need to allocate this anew every single iteration. It can be created with size of `unoriented_edges.len()` before the while loop.